### PR TITLE
Add op to convert numeric tensor to string faster

### PR DESCRIPTION
**Pull request by firejq**
When we want to convert a tensor of numeric value to string type, there is `tf.as_string` available for us. However, its performance will become very poor along with the input tensor swelling, which is caused due to its implement by vsnprintf in my opinion. vsnprintf indeed provides a quite rich formatting features but it turns into a encumbrance for the case that we only want to convert number to string.

Therefore, I proposed this pr trying to provide a high-performance and simple conversion method from numeric value to string type. **Performance improvement is the original intention of this pr**.

### Benchmark

I have made benchmarks against this new op and `tf.as_string` separately. For each method, I have tested both the case of float type and integer type, and the value of each element of input tensor is a random number from `-100,000` to `100,000`. For each input type, I have set up multiple gradually increasing inputs, the first of which is to measure the fixed overhead of the operation itself. In addition to that, each testing case is repeated 20 times and the average is taken as the result. 

*System environment: Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz*

The following benchmarks result (ms) is for input of float type:

| tensor shape     | tf.number_to_string | tf.as_string        |
| ---------------- | ------------------- | ------------------- |
| (0)              | 0.453  | 0.359  |
| (10,10)          | 0.458  | 0.403 |
| (100,100)        | 1.413   | 1.705  |
| (1,000, 100)     | 10.764    | 14.164  |
| (10,000, 100)    | 314.907  | 1025.248  |
| (100,000, 100)   | 2904.230  | 10173.129  |
| (1,000,000, 100) | 28880.552  |  101203.889  |

The following benchmarks result (ms) is for input of integer type:

| tensor shape     | tf.number_to_string | tf.as_string       |
| ---------------- | ------------------- | ------------------ |
| (0)              | 0.437  | 0.382 |
| (10,10)          | 0.438  | 0.389 |
| (100,100)        | 1.427  | 1.429 |
| (1,000, 100)     | 9.590   | 9.814  |
| (10,000, 100)    | 225.927  | 342.106   |
| (100,000, 100)   | 2119.049   | 3255.322  |
| (1,000,000, 100) | 21059.845   | 32273.443 |

From the test results, we can see that since the fixed overhead of `tf.number_to_string` is slightly larger than that of `tf.as_string`, when the input size is small, the performance of former is slightly lower (but personally this should be an acceptable gap, right?). **Along with the input size becoming larger (maybe `(100,100)` is the watershed), `tf.number_to_string` gradually shows a quite significant advantage in performance**. Among them, the performance of integer conversion has been improved x1.5, and x3.5 for floating point conversion.

Related discussion #45406 . Could you please have a look? @vnvo2409

### CODE REVIEW

File: tensorflow/core/kernels/number_to_string_op.cc
Comment (vnvo2409) [inline tensorflow/core/kernels/number_to_string_op.cc:1]: /s/2015/2020

File: third_party/fmtlib/BUILD
Comment (vnvo2409) [inline third_party/fmtlib/BUILD:18]: new line

**Review by vnghia — COMMENTED** [review]
Thank you for your PR. I am not a maintainer so let's wait for someone else to review this PR. In the meantime, in order to convince the maintainers, you should:
- add a benchmark between `tf.as_string` and `tf.number_to_string`.
- add some tests ( probably not much since `fmt` are already well-tested but at least some simple tests. )
- suggest some benefits for the codebase. e.g if we replace `tf.as_string` with `tf.number_to_string` in a file or in a model, the performance will increase a lot.

**Review by mdanatg — CHANGES_REQUESTED** [review]
Please include unit tests. Be sure to cover RaggedTensor if you intend to cover the op with add_dispatch_support.

File: tensorflow/workspace.bzl
Comment (mdanatg) [inline tensorflow/workspace.bzl:188]: We try to minimize the use of external dependencies. Any chance this job could be done by a standard library?

File: tensorflow/python/ops/string_ops.py
Comment (mdanatg) [inline tensorflow/python/ops/string_ops.py:498]: Iv the v1 name is the same as the main one, then you can omit the `v1=` arg.

File: tensorflow/python/kernel_tests/number_to_string_op_test.py
Comment (mdanatg) [inline tensorflow/python/kernel_tests/number_to_string_op_test.py:33]: This is great. Please also include a test or two with the following inputs:

`constant_op.constant`
`ragged_tensor.RaggedTensor` (from tensorflow.python.ops.ragged import ragged_tensor)

We also want to see what happens for non-numeric inputs, for example a `Variable` and a `TensorArray` (we expect an error).

File: tensorflow/core/kernels/number_to_string_op.cc
Comment (firejq) [reply tensorflow/core/kernels/number_to_string_op.cc]: Have done.

File: third_party/fmtlib/BUILD
Comment (firejq) [reply third_party/fmtlib/BUILD]: Have done.

File: tensorflow/workspace.bzl
Comment (mihaimaruseac) [reply tensorflow/workspace.bzl:188]: We have utilities in `strings::Printf` that should do this.

**Comment by firejq** [comment]
> Thank you for your PR. I am not a maintainer so let's wait for someone else to review this PR. In the meantime, in order to convince the maintainers, you should:
> 
> * add a benchmark between `tf.as_string` and `tf.number_to_string`.
> * add some tests ( probably not much since `fmt` are already well-tested but at least some simple tests. )
> * suggest some benefits for the codebase. e.g if we replace `tf.as_string` with `tf.number_to_string` in a file or in a model, the performance will increase a lot.

@vnvo2409 Thanks for your valuable suggestions. Sorry for the vagueness of my description. I have reorganized the description of pr and added performance benchmarks [here](https://github.com/tensorflow/tensorflow/pull/45466#issue-534189832).

File: tensorflow/workspace.bzl
Comment (firejq) [reply tensorflow/workspace.bzl:188]: I also noticed from the implementation of `tf.as_string` that the `strings::Printf` method can achieve exactly the same function. However, in my training scenario, because my model contains a large number of string conversions, this method has become a bottleneck for further performance breakthroughs. Considering that the performance of this implementation (i.e. `vsnprintf` in fact) is relatively poor while input swelling, and according to this [benchmark](http://www.zverovich.net/2020/06/13/fast-int-to-string-revisited.html), **`fmt::to_string` is 4x faster than `sprintf`**, so I introduced `fmt::to_string` and submitted this pr. 

P.S. I have reorganized the relevant instructions and benchmarks [here](https://github.com/tensorflow/tensorflow/pull/45466#issue-534189832), could you please have a look at it again?

Comment (mihaimaruseac) [reply tensorflow/workspace.bzl:188]: Maybe we can learn something from `fmt`'s implementation. The issue is that adding new dependencies might incur larger costs than the performance gain for a very infrequent scenario.

Comment (firejq) [reply tensorflow/workspace.bzl:188]: Instead of using `fmt` as a new third-party dependency, if we can migrate the implementation code of `fmt::to_string` to tensorflow (e.g. number.cc) alone, may it be an acceptable solution? Because in fact I only used one function in that library. WDYT? @mihaimaruseac

Comment (mihaimaruseac) [reply tensorflow/workspace.bzl:188]: That could work.

Alternatively, if you want to ue `fmt::to_string` there is the option of making this a custom op that you load when needed (from outside this repo)

**Comment by gbaned** [comment]
@firejq  Can you please check @mdanatg's comments and keep us posted ? Thanks!

**Comment by firejq** [comment]
> Can you please check @mdanatg's comments and keep us posted ? Thanks!

@gbaned Sorry for the late reply. In accordance with the amendments proposed by @mihaimaruseac [here](https://github.com/tensorflow/tensorflow/pull/45466#discussion_r539530149), I need to make considerable changes to the existing implementation to get rid of the dependence on third-party libraries `fmt`. As I have other tighter work on hand recently, I am going to close this pr first. If there are new developments in the future I will reopen it.

**Pull request closed without merging**
